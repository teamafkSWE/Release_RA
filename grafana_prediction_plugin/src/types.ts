export class Predictor {
    private readonly _name: string;
    private readonly _value: number;

    constructor(name: string, value: number) {
        this._name = name;
        this._value = value;
    }


    get name(): string {
        return this._name;
    }

    get value(): number {
        return this._value;
    }
}

export class Connection {

    private _id: string
    private _name: string
    private _links: { predictor: string, query: string }[]


    constructor(id: string, name: string, links: { predictor: string; query: string }[]) {
        this._id = id;
        this._name = name;
        this._links = links;
    }


    get id(): string {
        return this._id;
    }

    get name(): string {
        return this._name;
    }

    get links(): { predictor: string; query: string }[] {
        return this._links;
    }

    set id(value: string) {
        this._id = value;
    }

    set name(value: string) {
        this._name = value;
    }

    set links(value: { predictor: string; query: string }[]) {
        this._links = value;
    }
}

export class Datasource {
    private readonly _id: string
    private readonly _database: string;
    private readonly _name: string
    private readonly _url: URL
    private readonly _user: string
    private readonly _password: string


    constructor(id: string, database: string, name: string, url: string, user: string, password: string) {
        this._id = id.toString();
        this._database = database;
        this._name = name;
        this._url = new URL(url);
        this._user = user
        this._password = password
    }

    get user(): string {
        return this._user;
    }

    get password(): string {
        return this._password;
    }

    get id(): string {
        return this._id;
    }

    get database(): string {
        return this._database;
    }

    get name(): string {
        return this._name;
    }

    get url(): URL {
        return this._url;
    }
}