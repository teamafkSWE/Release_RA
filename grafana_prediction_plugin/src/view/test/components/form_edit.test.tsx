import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import {cleanup, render} from "@testing-library/react";
import Renderer from 'react-test-renderer'
import FormEdit from "../../components/form_edit";
import Controller from "../../../controller/controller";
import {Connection} from "../../../types";
import {AppEvents} from "@grafana/data";

class EmitterMock {
    emit = jest.fn()
}

const mockGetConnection = jest.fn().mockImplementation(() => {
    return [new Connection('1', 'x', [{query: 'cpu', predictor: 'y'}])]
})
const mockEditConnection = jest.fn()


describe('Form edit', () => {
    const queries = [{name: 'xy', fields: [], length: 0}, {name: 'yx', fields: [], length: 0}]
    const close = jest.fn()
    const controller = Controller.requireController(1)
    controller.getConnections = mockGetConnection
    controller.editConnection = mockEditConnection
    const emitter = new EmitterMock()
    beforeEach(() => {
        close.mockClear()
        emitter.emit.mockClear()
        mockGetConnection.mockClear()
        mockEditConnection.mockClear()
    })

    afterEach(() => {
        cleanup()
    })

    afterAll(() => Controller.unloadController(1))

    test('render', () => {
        const {asFragment,} = render(<FormEdit queries={queries} closeEdit={close} idConnection={'1'} controller={controller} emitter={emitter}/>)
        expect(asFragment()).toMatchSnapshot()
    })

    test('update name', () => {
        const instance: any = Renderer.create(<FormEdit queries={queries} closeEdit={close} idConnection={'1'} controller={controller} emitter={emitter}/>).getInstance()
        expect(instance.connectionName).toBe('x')
        instance.updateName({target: {value: 'xA'}})
        expect(instance.connectionName).toBe('xA')
    })

    test('update links', () => {
        const instance: any = Renderer.create(<FormEdit queries={queries} closeEdit={close} idConnection={'1'} controller={controller} emitter={emitter}/>).getInstance()
        expect(instance.connectionLinks).toStrictEqual([{query: 'cpu', predictor: 'y'}])
        instance.updateLinks({target: {value: 'gpu', id: 'y'}})
        expect(instance.connectionLinks).toStrictEqual([{query: 'gpu', predictor: 'y'}])
    })

    test('update connection empty name', () => {
        const instance: any = Renderer.create(<FormEdit queries={queries} closeEdit={close} idConnection={'1'} controller={controller} emitter={emitter}/>).getInstance()
        instance.connectionName = ''
        instance.updateConnection()
        expect(emitter.emit).toBeCalledTimes(1)
        expect(emitter.emit).toBeCalledWith(AppEvents.alertWarning, ["Enter a name for the connection:"])
    })

    test('update connection not empty name', () => {
        const instance: any = Renderer.create(<FormEdit queries={queries} closeEdit={close} idConnection={'1'} controller={controller} emitter={emitter}/>).getInstance()
        instance.updateConnection()
        expect(controller.editConnection).toBeCalledTimes(1)
        expect(controller.editConnection).toBeCalledWith('1', {name: 'x', list: [{query: 'cpu', predictor: 'y'}]})
        expect(close).toBeCalledTimes(1)
        expect(emitter.emit).toBeCalledTimes(1)
        expect(emitter.emit).toBeCalledWith(AppEvents.alertSuccess, ["Connection changed."])
    })
})