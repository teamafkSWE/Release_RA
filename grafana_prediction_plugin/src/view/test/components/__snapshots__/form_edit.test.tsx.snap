// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form edit render 1`] = `
<DocumentFragment>
  <div
    class="panel-options-group"
  >
    <div
      class="panel-options-group__header"
    >
      <span
        class="panel-options-group__title"
      >
        Edit connection
      </span>
    </div>
    <div
      class="panel-options-group__body"
    >
      <div
        class="css-dj00ry"
        style="width: 100%;"
      >
        <div
          class="css-186j473"
        >
          <label
            for="nome_collegamento"
          >
            Connection name:
          </label>
        </div>
        <div
          class="css-186j473"
        >
          <input
            id="nome_collegamento"
            placeholder="nome"
            style="width: 100%;"
            type="text"
          />
        </div>
        <div
          class="css-186j473"
        >
          <div
            style="display: flex; justify-content: space-between; margin-top: 0.8rem;"
          >
            <label
              for="y"
              style="align-self: center;"
            >
              y:
            </label>
            <select
              id="y"
              style="margin: 10px;"
            >
              <option
                value="xy"
              >
                xy
              </option>
              <option
                value="yx"
              >
                yx
              </option>
            </select>
          </div>
        </div>
        <div
          class="css-186j473"
        >
          <div
            style="display: flex; justify-content: space-evenly;"
          >
            <button
              class="css-1ihbihm-button"
            >
              <span
                class="css-1beih13"
              >
                <span>
                  Save
                </span>
              </span>
            </button>
            <button
              class="css-1ihbihm-button"
            >
              <span
                class="css-1beih13"
              >
                <span>
                  Cancel
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
